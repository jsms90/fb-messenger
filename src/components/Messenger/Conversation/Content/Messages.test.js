import React from 'react'
import { shallow, mount } from 'enzyme'
import { MemoryRouter as Router } from 'react-router-dom'
import { Provider } from 'react-redux'
import waitForExpect from 'wait-for-expect'

import configureStore from '../../../../store'
import ComposedMessages, { Messages, MessageBox, Message } from './Messages'

describe('<Messages />', () => {
  it(`should send a message (unit test)`, async () => {
    const username = "semra"
    const conversation = [{from: 'jen', to: 'semra', message: 'this is sooooo fun'}]
    const handleMessageReceipt = jest.fn()
    const fakeApi = {
      sendMessage: jest.fn()
    }
    const wrapper = shallow(<Messages username={username} receiveMessage={handleMessageReceipt} api={fakeApi} conversation={conversation}/>)
    
    await wrapper.find('button').simulate('click')
    expect(wrapper).toMatchSnapshot()
    expect(handleMessageReceipt).toHaveBeenCalled()
    expect(fakeApi.sendMessage).toHaveBeenCalled()

    // 1. shallow the <Messages /> component
    // You can use console.log(wrapper.debug()) to console.log the component that you are testing

    // 2  Mock the api. Hint, the api functions are passed as a defaultProp (look at the bottom of Messages.js),
    // you can override that prop by doing <Messages api={my_mocked_api_object} />

    // 3. Find the button -> you have an example here http://airbnb.io/enzyme/#shallow-rendering

    // 4. Click on the button -> you have an example here http://airbnb.io/enzyme/#shallow-rendering

    // 5. Assert the 'message was sent' -> 
    //      You can use toHaveBeenCalled on the my_mocked_api_object you passed. 
    //      toHaveBeenCalled needs a mock function https://jestjs.io/docs/en/mock-functions, is your sendMessage a mock?
    //      You have an example here http://airbnb.io/enzyme/#shallow-rendering heads-up!
    //      Enzyme expectations are not camel case,
    //      Jest expectations are camel case (for when you copy&paste :)

    // Final questions:
    // - Is this black-box testing or white-box testing?
    // - If I remove Redux from my application and put all the state in React, do I need to update this test?
    // - What's your level of confidence that the user will be able to send a message?
  })

  it(`should send a message (integration test)`, async () => {
    // 1. shallow or mount the <Messages /> component ?
    //    A) which component, Messages or ComposedMessages?
    //    B) If you mount the component then all the children are rendered. Hint: you need to provide a store.
    
    // 2  Mock the api. Hint, the api functions are passed as a defaultProp (look at the bottom of Messages.js),
    // you can override that prop by doing <Messages api={my_mocked_api_object} />
    const store = configureStore()

    const username = "semra"
    const conversation = [{from: 'jen', to: 'semra', message: 'this is sooooo fun'}]
    const handleMessageReceipt = jest.fn(message => Promise.resolve({id: 1, ...message}))
    const fakeApi = {
      sendMessage: jest.fn(message => Promise.resolve({id: 1, ...message}))
    }
    
    const wrapper = mount(
      <Provider store={store}>
        <ComposedMessages
          username={username}
          receiveMessage={handleMessageReceipt}
          api={fakeApi}
          conversation={conversation}
        />
      </Provider>
    )

    // 3. Add some text to the input
    expect(wrapper.find(Message).length).toEqual(0)
    wrapper.find(MessageBox).props().onChange({ target: { value: 'hi!' }})
    
    // 4. Find the button -> you have an example here http://airbnb.io/enzyme/#shallow-rendering
    const button = wrapper.find('button')
    
    // 5. Click on the button -> you have an example here http://airbnb.io/enzyme/#shallow-rendering
    // Heads-up! you need to use await on the click button. Or even better use https://www.npmjs.com/package/wait-for-expect
    button.simulate('click')
    
    // 6. You need to update the rendered component using http://airbnb.io/enzyme/docs/api/ShallowWrapper/update.html
    
    // 7. Assert the 'message was sent' -> you can just validate the message you sent is on the Messages list
    await waitForExpect(() => {
      wrapper.update()
      expect(wrapper.find(Message).length).toBe(1)
      expect(wrapper.find(Message).text()).toEqual('hi!')
    })

    wrapper.find(MessageBox).props().onChange({ target: { value: 'what\'s up?' }})
    button.simulate('click')
    await waitForExpect(() => {
      wrapper.update()
      expect(wrapper.find(Message).length).toBe(2)
      expect(wrapper.find(Message).at(1).text()).toEqual('what\'s up?')
    })

    // Final questions:
    // - Is this black-box testing or white-box testing?
    // - If I remove Redux from my application and put all the state in React, do I need to update this test?
    // - What's your level of confidence that the user will be able to send a message?
  })
})
