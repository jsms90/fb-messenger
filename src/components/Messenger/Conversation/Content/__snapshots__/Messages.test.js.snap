// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Messages /> should send a message (unit test) 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Messages
    api={
      Object {
        "sendMessage": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "message": "",
                "to": "semra",
              },
            ],
          ],
        },
      }
    }
    conversation={
      Array [
        Object {
          "from": "jen",
          "message": "this is sooooo fun",
          "to": "semra",
        },
      ]
    }
    receiveMessage={
      [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
      }
    }
    username="semra"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <styled.div>
          <styled.div
            from="received"
          >
            <Message
              from="received"
            >
              this is sooooo fun
            </Message>
          </styled.div>
        </styled.div>,
        <styled.div>
          <styled.input
            onChange={[Function]}
            placeholder="Type your message..."
            type="text"
            value=""
          />
          <button
            onClick={[Function]}
          >
            Send
          </button>
        </styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <styled.div
              from="received"
            >
              <Message
                from="received"
              >
                this is sooooo fun
              </Message>
            </styled.div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "children": Array [
                false,
                <Message
                  from="received"
                >
                  this is sooooo fun
                </Message>,
                false,
              ],
              "from": "received",
            },
            "ref": null,
            "rendered": Array [
              false,
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "this is sooooo fun",
                  "from": "received",
                },
                "ref": null,
                "rendered": "this is sooooo fun",
                "type": [Function],
              },
              false,
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <styled.input
              onChange={[Function]}
              placeholder="Type your message..."
              type="text"
              value=""
            />,
            <button
              onClick={[Function]}
            >
              Send
            </button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onChange": [Function],
              "placeholder": "Type your message...",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Send",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Send",
            "type": "button",
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <styled.div>
            <styled.div
              from="received"
            >
              <Message
                from="received"
              >
                this is sooooo fun
              </Message>
            </styled.div>
          </styled.div>,
          <styled.div>
            <styled.input
              onChange={[Function]}
              placeholder="Type your message..."
              type="text"
              value=""
            />
            <button
              onClick={[Function]}
            >
              Send
            </button>
          </styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <styled.div
                from="received"
              >
                <Message
                  from="received"
                >
                  this is sooooo fun
                </Message>
              </styled.div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  false,
                  <Message
                    from="received"
                  >
                    this is sooooo fun
                  </Message>,
                  false,
                ],
                "from": "received",
              },
              "ref": null,
              "rendered": Array [
                false,
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "this is sooooo fun",
                    "from": "received",
                  },
                  "ref": null,
                  "rendered": "this is sooooo fun",
                  "type": [Function],
                },
                false,
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <styled.input
                onChange={[Function]}
                placeholder="Type your message..."
                type="text"
                value=""
              />,
              <button
                onClick={[Function]}
              >
                Send
              </button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "onChange": [Function],
                "placeholder": "Type your message...",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Send",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Send",
              "type": "button",
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
